### PR TITLE
Add base loop task execotor and http template parser

### LINKED ISSUE #11103: [Feature][Task] Add a new abstract task-AbstractLoopStateTask
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

# Background
Right now, there exist some tasks, that need to call a third-party platform, and query the task instance status from the third-party platform until the task instance is finished (success/failed).

image

# Class design
image

We can create a base class for this kink of task
```java
 /**
 * This class is the base class for all loop task type.
 * <p>
 * The loop task type means, we will submit a task, and loop the task status until the task is finished.
 */
public abstract class BaseLoopTaskExecutor<SubmitTaskMethodT extends LoopTaskSubmitTaskMethodDefinition, QueryTaskMethodT extends LoopTaskQueryStatusMethodDefinition>
    extends AbstractTaskExecutor {

    protected final LoopTaskDefinition<SubmitTaskMethodT, QueryTaskMethodT> loopTaskDefinition;
    // init this parameters in init method(todo: we need to refactor the init method)
    protected AbstractParameters loopTaskParameters;

    protected BaseLoopTaskExecutor(@NonNull TaskExecutionContext taskExecutionContext,
                                   @NonNull LoopTaskDefinition<SubmitTaskMethodT, QueryTaskMethodT> loopTaskDefinition) {
        super(taskExecutionContext);
        this.loopTaskDefinition = loopTaskDefinition;
    }

    @Override
    public void handle() throws TaskException {
        try {
            LoopTaskSubmitTaskMethodDefinition submitTaskMethod = loopTaskDefinition.getSubmitTaskMethod();
            LoopTaskQueryStatusMethodDefinition queryTaskStateMethod = loopTaskDefinition.getQueryTaskStateMethod();
            long loopInterval = queryTaskStateMethod.getTaskInstanceStatusQueryInterval().toMillis();

            LoopTaskInstanceInfo loopTaskInstanceInfo = submitTaskMethod.submitLoopTaskInstance();
            this.appIds = loopTaskInstanceInfo.getTaskInstanceId();
            while (!queryTaskStateMethod.queryTaskInstanceStatus(loopTaskInstanceInfo).isFinished()) {
                Thread.sleep(loopInterval);
            }
        } catch (InterruptedException e) {
            setExitStatusCode(TaskConstants.EXIT_CODE_FAILURE);
            logger.error("The current loop thread has been interrupted", e);
            Thread.currentThread().interrupt();
            throw new TaskException("The current loop thread has been interrupted");
        } catch (Exception ex) {
            setExitStatusCode(TaskConstants.EXIT_CODE_FAILURE);
            logger.error("Loop task execute error", ex);
            throw new TaskException("Loop task execute error", ex);
        }
    }
}
```

If someone wants to implement this kind of task, he only needs to provide a LoopTaskDefinition.

In deed, we can provide the default implementation for HTTP call, if someone submitTask/ queryStatus by HTTP call, he can just provide a YAML config, we will translate these to method (But I think only a few people will use this).

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by ruanwenjun**
## Purpose of the pull request

close #11103

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

### CODE REVIEW

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/loop/BaseLoopTaskExecutor.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    @Override
+    public void handle() throws TaskException {
+        try {
+            final long loopInterval = getTaskInstanceStatusQueryInterval().toMillis();
+            loopTaskInstanceInfo = submitLoopTask();
+            this.appIds = loopTaskInstanceInfo.getTaskInstanceId();
+            while (!cancel
+                && !RetryUtils.retryFunction(() -> queryTaskInstanceStatus(loopTaskInstanceInfo).isFinished())) {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/loop/BaseLoopTaskExecutor.java:58]: please add some comments.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/loop/template/http/method/HttpLoopTaskCancelTaskMethodDefinition.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                                                  String httpMethodType,
+                                                  String dataType,
+                                                  Map<String, Object> requestParams,
+                                                  Map<String, Object> requestBody) {
+        super(url, httpMethodType, dataType, requestParams, requestBody);
+    }
+
+    @Override
+    public void cancelTaskInstance(LoopTaskInstanceInfo loopTaskInstanceInfo) {
+        if (requestParams != null) {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/loop/template/http/method/HttpLoopTaskCancelTaskMethodDefinition.java:50]: It seems some same code.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/OkHttpUtils.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        HttpUrl.Builder urlBuilder = httpUrl.newBuilder();
+        if (requestParams != null) {
+            for (Map.Entry<String, Object> entry : requestParams.entrySet()) {
+                urlBuilder.addQueryParameter(entry.getKey(), entry.getValue().toString());
+            }
+        }
+        Request request = new Request.Builder().url(urlBuilder.build()).build();
+        try (Response response = CLIENT.newCall(request).execute()) {
+            if (response.code() != 200 || response.body() == null) {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/OkHttpUtils.java]:
```suggestion
            if (response.code() != HttpStatus.OK || response.body() == null) {
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/OkHttpUtils.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                urlBuilder.addQueryParameter(entry.getKey(), entry.getValue().toString());
+            }
+        }
+        Request.Builder requestBuilder = new Request.Builder().url(urlBuilder.build());
+        if (requestBodyMap != null) {
+            requestBuilder = requestBuilder.post(RequestBody.create(MediaType.parse("application/json"),
+                                                                    JSONUtils.toJsonString(requestBodyMap)));
+        }
+        try (Response response = CLIENT.newCall(requestBuilder.build()).execute()) {
+            if (response.code() != 200 || response.body() == null) {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/OkHttpUtils.java]:
```suggestion
            if (response.code() != HttpStatus.OK || response.body() == null) {
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/RetryUtils.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        while (true) {
+            try {
+                return supplier.get();
+            } catch (Exception ex) {
+                if (retryCount == retryPolicy.getMaxRetryTimes()) {
+                    throw ex;
+                }
+                retryCount++;
+                try {
+                    Thread.sleep(retryInterval);
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/RetryUtils.java:54]: Maybe we should tell users of the thread sleep when call this method.

Comment (ruanwenjun) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/RetryUtils.java:54]: Done

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/OkHttpUtils.java (hunk 3 above)
Comment (ruanwenjun) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/OkHttpUtils.java]: Done

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/OkHttpUtils.java (hunk 4 above)
Comment (ruanwenjun) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/utils/OkHttpUtils.java]: Done

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/loop/BaseLoopTaskExecutor.java (hunk 1 above)
Comment (ruanwenjun) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/loop/BaseLoopTaskExecutor.java:58]: Done

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/loop/template/http/method/HttpLoopTaskCancelTaskMethodDefinition.java (hunk 2 above)
Comment (ruanwenjun) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/loop/template/http/method/HttpLoopTaskCancelTaskMethodDefinition.java:50]: Yes, right now there are some same code in the queryStatus/cancelTask, but these two methods will be different :) they may inject parameter from other place in the future.

**Comment by zhuxt2015** [comment]
I have three small points:

1. The use of the mode is slightly complicated, requiring a third-party platform to provide an interface without user authentication
2. If the number of tasks is large, it may slow down the performance of the worker.
3. I think the better approach is: When starting a workflow, a unique ID is returned as the ID of the workflow instance, third-party platform use the instance ID to query  status of workflow instance and tasks.

@ruanwenjun  WDYT?

**Comment by ruanwenjun** [comment]
> I have three small points:
> 
> 1. The use of the mode is slightly complicated, requiring a third-party platform to provide an interface without user authentication

In fact, there exist these kinds of tasks, ds is as a scheduler trigger to submit tasks to these platforms.

> 2. If the number of tasks is large, it may slow down the performance of the worker.

Yes, we can use single thread pool to loop the status in the future, but I think this will not heavier than process.

> 3. When starting a workflow, a unique ID is returned as the ID of the workflow instance, third-party platform use the instance ID to query  status of workflow instance and tasks.

Third-party platform can call API-Server to query the status, this PR is used to provide a base task for `query status task` like EMR.

**Review by caishunfeng — APPROVED** [review]
LGTM

**Pull request merged**